lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2.rb

```diff
@@ -0,0 +1,318 @@
+  # supplied definition.
+  #
+  # Returns nil if the index does not exist
+  #
+  # Note - This is a 1 based array (not zero)
+  #
+  def self.trigger(definition, index)
+    result = nil
+
+    begin
+      result = definition.Triggers.Item(index)
+    rescue WIN32OLERuntimeError => err
+      # E_INVALIDARG 0x80070057 from # https://msdn.microsoft.com/en-us/library/windows/desktop/aa378137%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
+      if err.message =~ /80070057/m
+        result = nil
+      else
+        raise
+      end
+    end
+
+    result
+  end
+
+  def self.append_new_trigger(definition, trigger_type)
+    definition.Triggers.create(trigger_type)
+  end
+
+  # Deletes the trigger at the specified index.
+  #
+  def self.delete_trigger(definition, index)
+    definition.Triggers.Remove(index)
+
+    index
+  end
+
+  # Helpers
+
+  # From https://msdn.microsoft.com/en-us/library/windows/desktop/aa381850(v=vs.85).aspx
+  # https://en.wikipedia.org/wiki/ISO_8601#Durations
+  #
+  # The format for this string is PnYnMnDTnHnMnS, where nY is the number of years, nM is the number of months,
+  # nD is the number of days, 'T' is the date/time separator, nH is the number of hours, nM is the number of minutes,
+  # and nS is the number of seconds (for example, PT5M specifies 5 minutes and P1M4DT2H5M specifies one month,
+  # four days, two hours, and five minutes)
+  def self.duration_to_hash(duration)
+    regex = /^P((?'year'\d+)Y)?((?'month'\d+)M)?((?'day'\d+)D)?(T((?'hour'\d+)H)?((?'minute'\d+)M)?((?'second'\d+)S)?)?$/
+
+    matches = regex.match(duration)
+    return nil if matches.nil?
+
+    {
+      :year => matches['year'],
+      :month => matches['month'],
+      :day => matches['day'],
+      :minute => matches['minute'],
+      :hour => matches['hour'],
+      :second => matches['second'],
+    }
+  end
+
+  # Converts a hash table describing year, month, day etc. into a timelimit string as per
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa381850(v=vs.85).aspx
+  # https://en.wikipedia.org/wiki/ISO_8601#Durations
+  # returns PT0S if there is nothing set.
+  def self.hash_to_duration(hash)
+    duration = 'P'
+    duration += hash[:year].to_s + 'Y' unless hash[:year].nil? || hash[:year].zero?
+    duration += hash[:month].to_s + 'M' unless hash[:month].nil? || hash[:month].zero?
+    duration += hash[:day].to_s + 'D' unless hash[:day].nil? || hash[:day].zero?
+    duration += 'T'
+    duration += hash[:hour].to_s + 'H' unless hash[:hour].nil? || hash[:hour].zero?
+    duration += hash[:minute].to_s + 'M' unless hash[:minute].nil? || hash[:minute].zero?
+    duration += hash[:second].to_s + 'S' unless hash[:second].nil? || hash[:second].zero?
+
+    duration == 'PT' ? 'PT0S' : duration
+  end
+
+  # Private methods
+  def self.task_service
+    if @@service_object.nil?
+      @@service_object = WIN32OLE.new('Schedule.Service')
```

Comment:
Will an effective singleton here cause us any problems? Are these expensive to create?

Reply:
Given this is in a module I'm not sure why I used `@@`.  Should probably just use `@`

Creating OLE objects is "expensive" but relative to a puppet run, probably not.  It just seemed a waste to continuously create and destroy that object, and just keep it around.

Reply:
For now I've changed it to `@`

Reply:
Maybe you started with `@@` when this code was a class / before changing to module?

I was mostly just concerned with understanding the appropriate usage pattern for that COM component.... For instance, I was wondering if it's possible to get the object into a state where a failure in one resource ends up putting that object into a state that makes it unable to service other resources in the same run, etc. I would hope that would not be the case, but given the v1 weirdness, thought it was worth asking the question.

Reply:
In theory yes. It could not recover if the object was in a bad state, whereas if I created and connected on every invocation then it would auto-correct

Reply:
However that service object shouldn't really store state between calls, it's all about;
1. call `get` task
2. modify the object returned using its methods etc.
3. call `set` task

I'm happy to change it to create a new object on each call.  I suspect the actual time difference is minimal as specs use the code in a different way than a puppet resource.